packages/backend/permissions.go

```diff
@@ -0,0 +1,210 @@
+package main
+import (
+    pb "github.com/canonical/desktop-security-center/packages/proto"
+    epb "google.golang.org/protobuf/types/known/emptypb"
+    wpb "google.golang.org/protobuf/types/known/wrapperspb"
+    "context"
+    "github.com/tidwall/gjson"
+    "io"
+    "bytes"
+    "log"
+    "net/http"
+    "fmt"
+    "google.golang.org/grpc/codes"
+    "google.golang.org/grpc/status"
+)
+
+const (
+  rulesApi = "http://localhost/v2/interfaces/requests/rules"
+  confApi = "http://localhost/v2/snaps/system/conf"
+  removeBody = `{"action":"remove"}`
+  promptBody = `{"experimental.apparmor-prompting":%s}`
+  sep = ":"
+)
+var empty = new(epb.Empty)
+var snapPathId = map[string]string{}
+var authHeader = map[string]string{"X-Allow-Interaction": "true"}
+
+type HttpClient interface {
+    Do(req *http.Request) (*http.Response, error)
+}
+
+type PermissionServer struct {
+    pb.UnimplementedPermissionServer
+    client HttpClient
+}
+
+func NewPermissionServer(c HttpClient) (*PermissionServer) {
+    /* Copy paste from https://stackoverflow.com/a/59665098 */
+    s := &PermissionServer{
+        client: c,
+    }
+    return s
+}
+
+func makeRestReq(client HttpClient, kind string, headers map[string]string, where string, reqBody io.Reader) (string, error) {
+    req, err := http.NewRequest(kind, where, reqBody)
+    if err != nil {
+        log.Printf("Couldn't create %s request to %s.", kind, where)
+        return "", err
+    }
+    for k, val := range headers {
+        req.Header.Add(k, val)
+    }
+    res, err := client.Do(req)
+    if err != nil {
+        log.Printf("Couldn't send %s request to %s.", kind, where)
+        return "", err
+    }
+    defer res.Body.Close()
+    resBody, err := io.ReadAll(res.Body)
+    if err != nil {
+        log.Printf("Couldn't read response from %s request to %s.", kind, where)
+        return "", err
+    }
+    return string(resBody), nil
+}
+
+func validateApiResponse(res string) error {
+    if !gjson.Valid(res) {
```

Comment:
@matthew-hagemann I had a chat about this on a call with @nteodosio earlier: can you provide some references of how the JSON parsing works in `provd` for @nteodosio to take a look at please? :slightly_smiling_face:

Reply:
Sure thing!

@nteodosio  So here is an example from the service that communicated with `pro`

[Here](https://github.com/canonical/ubuntu-desktop-provision/blob/main/provd/internal/services/pro/pro.go#L54-L75) I define the structure of the json response the pro client sends back as a go struct with annotations.

And then you can quite easily [unmarshal](https://github.com/canonical/ubuntu-desktop-provision/blob/main/provd/internal/services/pro/pro.go#L175-L176)  the response and handle any errors, leaving you with a go struct you can work with.

Reply:
Not fun when the value can be either an array or a 'simple type'.

Reply:
E.g. the same end-point might return

```
{
  "result": {
    "message": "rule ID is not found"
  }
}
```
or
```
{
  "result": [
    {
      "id": "C7JHBW7E7Q7PO===",
      "timestamp": "2024-05-24T13:48:17.723465463Z",
      "user": 1000,
      "outcome": "allow",
      ...
    }
  ]
}
```